File: frontend/lib/models/order.dart
import 'user.dart';
import 'product.dart';

enum OrderStatus {
  pending,
  confirmed,
  pickedUp,
  inTransit,
  delivered,
  cancelled,
  disputed,
}

enum PaymentMethod { card, cash, payme, click }

enum PaymentStatus { pending, held, completed, refunded, failed }

class Order {
  final String id;
  final String orderNumber;
  final String buyerId;
  final String sellerId;
  final String? courierId;
  final String productId;
  final String? videoId;
  final int quantity;
  final double unitPrice;
  final double totalAmount;
  final double courierFee;
  final double platformCommission;
  final double sellerAmount;
  final String currency;
  final OrderStatus status;
  final PaymentMethod paymentMethod;
  final PaymentStatus paymentStatus;
  final String shippingAddress;
  final String shippingCity;
  final String shippingPhone;
  final String? buyerNote;
  final String? sellerQrCode;
  final String? courierQrCode;
  final String? deliveryCode;
  final DateTime? pickedUpAt;
  final DateTime? deliveredAt;
  final DateTime? cancelledAt;
  final String? cancelReason;
  final User? buyer;
  final User? seller;
  final User? courier;
  final Product? product;
  final DateTime? createdAt;

  Order({
    required this.id,
    required this.orderNumber,
    required this.buyerId,
    required this.sellerId,
    this.courierId,
    required this.productId,
    this.videoId,
    required this.quantity,
    required this.unitPrice,
    required this.totalAmount,
    required this.courierFee,
    required this.platformCommission,
    required this.sellerAmount,
    this.currency = 'UZS',
    required this.status,
    required this.paymentMethod,
    required this.paymentStatus,
    required this.shippingAddress,
    required this.shippingCity,
    required this.shippingPhone,
    this.buyerNote,
    this.sellerQrCode,
    this.courierQrCode,
    this.deliveryCode,
    this.pickedUpAt,
    this.deliveredAt,
    this.cancelledAt,
    this.cancelReason,
    this.buyer,
    this.seller,
    this.courier,
    this.product,
    this.createdAt,
  });

  bool get canCancel =>
      status == OrderStatus.pending || status == OrderStatus.confirmed;

  bool get isCompleted => status == OrderStatus.delivered;

  bool get isCancelled => status == OrderStatus.cancelled;

  String get statusText {
    switch (status) {
      case OrderStatus.pending:
        return 'Pending';
      case OrderStatus.confirmed:
        return 'Confirmed';
      case OrderStatus.pickedUp:
        return 'Picked Up';
      case OrderStatus.inTransit:
        return 'In Transit';
      case OrderStatus.delivered:
        return 'Delivered';
      case OrderStatus.cancelled:
        return 'Cancelled';
      case OrderStatus.disputed:
        return 'Disputed';
    }
  }

  factory Order.fromJson(Map<String, dynamic> json) {
    return Order(
      id: json['id'] as String,
      orderNumber: json['orderNumber'] as String,
      buyerId: json['buyerId'] as String,
      sellerId: json['sellerId'] as String,
      courierId: json['courierId'] as String?,
      productId: json['productId'] as String,
      videoId: json['videoId'] as String?,
      quantity: json['quantity'] as int,
      unitPrice: (json['unitPrice'] as num).toDouble(),
      totalAmount: (json['totalAmount'] as num).toDouble(),
      courierFee: (json['courierFee'] as num).toDouble(),
      platformCommission: (json['platformCommission'] as num).toDouble(),
      sellerAmount: (json['sellerAmount'] as num).toDouble(),
      currency: json['currency'] as String? ?? 'UZS',
      status: OrderStatus.values.firstWhere(
        (e) => e.name == json['status'],
        orElse: () => OrderStatus.pending,
      ),
      paymentMethod: PaymentMethod.values.firstWhere(
        (e) => e.name == json['paymentMethod'],
        orElse: () => PaymentMethod.cash,
      ),
      paymentStatus: PaymentStatus.values.firstWhere(
        (e) => e.name == json['paymentStatus'],
        orElse: () => PaymentStatus.pending,
      ),
      shippingAddress: json['shippingAddress'] as String,
      shippingCity: json['shippingCity'] as String,
      shippingPhone: json['shippingPhone'] as String,
      buyerNote: json['buyerNote'] as String?,
      sellerQrCode: json['sellerQrCode'] as String?,
      courierQrCode: json['courierQrCode'] as String?,
      deliveryCode: json['deliveryCode'] as String?,
      pickedUpAt: json['pickedUpAt'] != null
          ? DateTime.parse(json['pickedUpAt'] as String)
          : null,
      deliveredAt: json['deliveredAt'] != null
          ? DateTime.parse(json['deliveredAt'] as String)
          : null,
      cancelledAt: json['cancelledAt'] != null
          ? DateTime.parse(json['cancelledAt'] as String)
          : null,
      cancelReason: json['cancelReason'] as String?,
      buyer: json['buyer'] != null
          ? User.fromJson(json['buyer'] as Map<String, dynamic>)
          : null,
      seller: json['seller'] != null
          ? User.fromJson(json['seller'] as Map<String, dynamic>)
          : null,
      courier: json['courier'] != null
          ? User.fromJson(json['courier'] as Map<String, dynamic>)
          : null,
      product: json['product'] != null
          ? Product.fromJson(json['product'] as Map<String, dynamic>)
          : null,
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'] as String)
          : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'orderNumber': orderNumber,
      'buyerId': buyerId,
      'sellerId': sellerId,
      'courierId': courierId,
      'productId': productId,
      'videoId': videoId,
      'quantity': quantity,
      'unitPrice': unitPrice,
      'totalAmount': totalAmount,
      'courierFee': courierFee,
      'platformCommission': platformCommission,
      'sellerAmount': sellerAmount,
      'currency': currency,
      'status': status.name,
      'paymentMethod': paymentMethod.name,
      'paymentStatus': paymentStatus.name,
      'shippingAddress': shippingAddress,
      'shippingCity': shippingCity,
      'shippingPhone': shippingPhone,
      'buyerNote': buyerNote,
      'sellerQrCode': sellerQrCode,
      'courierQrCode': courierQrCode,
      'deliveryCode': deliveryCode,
      'pickedUpAt': pickedUpAt?.toIso8601String(),
      'deliveredAt': deliveredAt?.toIso8601String(),
      'cancelledAt': cancelledAt?.toIso8601String(),
      'cancelReason': cancelReason,
      'createdAt': createdAt?.toIso8601String(),
    };
  }
}

File: backend/src/models/Order.ts
import { DataTypes, Model, Optional } from 'sequelize';
import sequelize from '../config/database';
import { OrderStatus, PaymentMethod, PaymentStatus } from '../types';

interface OrderAttributes {
  id: string;
  orderNumber: string;
  buyerId: string;
  sellerId: string;
  courierId?: string;
  productId: string;
  videoId?: string;
  quantity: number;
  unitPrice: number;
  totalAmount: number;
  courierFee: number;
  platformCommission: number;
  sellerAmount: number;
  currency: string;
  status: OrderStatus;
  paymentMethod: PaymentMethod;
  paymentStatus: PaymentStatus;
  shippingAddress: string;
  shippingCity: string;
  shippingPhone: string;
  buyerNote?: string;
  sellerQrCode?: string;
  courierQrCode?: string;
  deliveryCode?: string;
  pickedUpAt?: Date;
  deliveredAt?: Date;
  cancelledAt?: Date;
  cancelReason?: string;
  createdAt?: Date;
  updatedAt?: Date;
}

interface OrderCreationAttributes extends Optional<OrderAttributes, 'id' | 'orderNumber' | 'courierId' | 'videoId' | 'buyerNote' | 'sellerQrCode' | 'courierQrCode' | 'deliveryCode' | 'pickedUpAt' | 'deliveredAt' | 'cancelledAt' | 'cancelReason' | 'createdAt' | 'updatedAt'> {}

class Order extends Model<OrderAttributes, OrderCreationAttributes> implements OrderAttributes {
  public id!: string;
  public orderNumber!: string;
  public buyerId!: string;
  public sellerId!: string;
  public courierId?: string;
  public productId!: string;
  public videoId?: string;
  public quantity!: number;
  public unitPrice!: number;
  public totalAmount!: number;
  public courierFee!: number;
  public platformCommission!: number;
  public sellerAmount!: number;
  public currency!: string;
  public status!: OrderStatus;
  public paymentMethod!: PaymentMethod;
  public paymentStatus!: PaymentStatus;
  public shippingAddress!: string;
  public shippingCity!: string;
  public shippingPhone!: string;
  public buyerNote?: string;
  public sellerQrCode?: string;
  public courierQrCode?: string;
  public deliveryCode?: string;
  public pickedUpAt?: Date;
  public deliveredAt?: Date;
  public cancelledAt?: Date;
  public cancelReason?: string;
  public readonly createdAt!: Date;
  public readonly updatedAt!: Date;
}

Order.init(
  {
    id: {
      type: DataTypes.UUID,
      defaultValue: DataTypes.UUIDV4,
      primaryKey: true,
    },
    orderNumber: {
      type: DataTypes.STRING(20),
      allowNull: false,
      unique: true,
    },
    buyerId: {
      type: DataTypes.UUID,
      allowNull: false,
      references: {
        model: 'users',
        key: 'id',
      },
    },
    sellerId: {
      type: DataTypes.UUID,
      allowNull: false,
      references: {
        model: 'users',
        key: 'id',
      },
    },
    courierId: {
      type: DataTypes.UUID,
      allowNull: true,
      references: {
        model: 'users',
        key: 'id',
      },
    },
    productId: {
      type: DataTypes.UUID,
      allowNull: false,
      references: {
        model: 'products',
        key: 'id',
      },
    },
    videoId: {
      type: DataTypes.UUID,
      allowNull: true,
      references: {
        model: 'videos',
        key: 'id',
      },
    },
    quantity: {
      type: DataTypes.INTEGER,
      allowNull: false,
      defaultValue: 1,
    },
    unitPrice: {
      type: DataTypes.DECIMAL(12, 2),
      allowNull: false,
    },
    totalAmount: {
      type: DataTypes.DECIMAL(12, 2),
      allowNull: false,
    },
    courierFee: {
      type: DataTypes.DECIMAL(12, 2),
      allowNull: false,
      defaultValue: 0,
    },
    platformCommission: {
      type: DataTypes.DECIMAL(12, 2),
      allowNull: false,
      defaultValue: 0,
    },
    sellerAmount: {
      type: DataTypes.DECIMAL(12, 2),
      allowNull: false,
    },
    currency: {
      type: DataTypes.STRING(3),
      defaultValue: 'UZS',
    },
    status: {
      type: DataTypes.ENUM(...Object.values(OrderStatus)),
      defaultValue: OrderStatus.PENDING,
    },
    paymentMethod: {
      type: DataTypes.ENUM(...Object.values(PaymentMethod)),
      allowNull: false,
    },
    paymentStatus: {
      type: DataTypes.ENUM(...Object.values(PaymentStatus)),
      defaultValue: PaymentStatus.PENDING,
    },
    shippingAddress: {
      type: DataTypes.TEXT,
      allowNull: false,
    },
    shippingCity: {
      type: DataTypes.STRING(100),
      allowNull: false,
    },
    shippingPhone: {
      type: DataTypes.STRING(20),
      allowNull: false,
    },
    buyerNote: {
      type: DataTypes.TEXT,
      allowNull: true,
    },
    sellerQrCode: {
      type: DataTypes.TEXT,
      allowNull: true,
    },
    courierQrCode: {
      type: DataTypes.TEXT,
      allowNull: true,
    },
    deliveryCode: {
      type: DataTypes.STRING(6),
      allowNull: true,
    },
    pickedUpAt: {
      type: DataTypes.DATE,
      allowNull: true,
    },
    deliveredAt: {
      type: DataTypes.DATE,
      allowNull: true,
    },
    cancelledAt: {
      type: DataTypes.DATE,
      allowNull: true,
    },
    cancelReason: {
      type: DataTypes.TEXT,
      allowNull: true,
    },
  },
  {
    sequelize,
    tableName: 'orders',
    modelName: 'Order',
    hooks: {
      beforeCreate: (order) => {
        const timestamp = Date.now().toString(36).toUpperCase();
        const random = Math.random().toString(36).substring(2, 6).toUpperCase();
        order.orderNumber = `GGM-${timestamp}-${random}`;
      },
    },
  }
);

export default Order;
